futures-channel/src/mpsc/mod.rs

```diff
@@ -820,10 +833,15 @@ impl<T> Stream for Receiver<T> {
     ) -> Poll<Option<T>> {
             // Try to read a message off of the message queue.
         match self.next_message() {
-            Poll::Ready(msg) => Poll::Ready(msg),
+            Poll::Ready(msg) => {
+                if msg.is_none() {
+                    self.inner = None;
+                }
+                Poll::Ready(msg)
+            },
             Poll::Pending => {
                 // There are no messages to read, in this case, park.
-                self.inner.recv_task.register(lw);
+                self.inner.as_ref().unwrap().recv_task.register(lw);
                 // Check queue again after parking to prevent race condition:
                 // a message could be added to the queue after previous `next_message`
                 // before `register` call.
```

Comment:
The below `self.next_message()` could return `Poll::Ready(None)` and then not update `self.inner`.

Relatedly tests are failing because `Drop` will call `next_message` at least once, even if `self.inner` is `None`; and `try_next` also calls it and I'm not sure what the best semantics for it would be.

Maybe `next_message` should return `Option<Poll<Option<T>>>` to make handling a post-termination call the callers job?

Reply:
I think it's correct for `try_next` to panic in this case as well-- I'll add a note to the docs, and fix the `Drop` impl. Thanks!